### PR TITLE
force_stride_order on fused_all_gather_matmul/fused_matmul_reduce_scatter's operands to avoid a copy due to layout transformation

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127682
* #127556
* #127455
* __->__ #127454

When performing fused_all_gather_matmul/fused_matmul_reduce_scatter and gather_dim/scatter_dim != 0, a copy of the lhs operand (A_shard/A) is needed for layout transformation.
This copy can be avoided if the lhs operand already has the following stride order:

    lhs.movedim(gather_dim, 0).contiguous().movedim(0, gather_dim).stride()

In `micro_pipeline_tp` passes, we enforce the lhs operand to have such stride order via `inductor_prims.force_stride_order`. This way if the lhs operand has a flexible layout, the copy is avoided.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

File: torch/distributed/_cuda_p2p/__init__.py [hunk 1]
```diff
@@ -404,6 +409,14 @@ def shard_consumer(shard: torch.Tensor, rank: int) -> None:
     return unflatten(ag_out), [unflatten(output) for output in outputs]
 
 
+def get_optimal_A_shard_stride_order_for_fused_all_gather_matmul(
```
Comment (Chillee) [inline torch/distributed/_cuda_p2p/__init__.py:412]: Can we refactor this to an utility that given a contiguous tensor, converts it to a tensor with a given permuted layout?

I think something along the lines of
```
def restride_tensor(x, perm):
    return x.permute(perm).contiguous().permute(inv_perm)
```
would work.

**Review by Chillee — CHANGES_REQUESTED** [review]

File: torch/distributed/_cuda_p2p/__init__.py (hunk 1 above)
Comment (yifuwang) [reply torch/distributed/_cuda_p2p/__init__.py:412]: Do I need the `needs_fixed_stride_order` tag on the operator for this approach to work?

Comment (Chillee) [reply torch/distributed/_cuda_p2p/__init__.py:412]: You can still do it within Inductor, just a suggestion on how to organize the code a bit nicer.

**Review by Chillee — APPROVED** [review]

**Pull request closed without merging**